polkadot/xcm/src/lib.rs

```diff
@@ -310,40 +339,21 @@ versioned_type! {
 	}
 }
 
-/// A single XCM message, together with its version code.
-#[derive(Derivative, Encode, Decode, TypeInfo)]
-#[derivative(Clone(bound = ""), Eq(bound = ""), PartialEq(bound = ""), Debug(bound = ""))]
-#[codec(encode_bound())]
-#[codec(decode_bound())]
-#[scale_info(bounds(), skip_type_params(RuntimeCall))]
-#[scale_info(replace_segment("staging_xcm", "xcm"))]
-pub enum VersionedXcm<RuntimeCall> {
-	#[codec(index = 3)]
-	V3(v3::Xcm<RuntimeCall>),
-	#[codec(index = 4)]
-	V4(v4::Xcm<RuntimeCall>),
-	#[codec(index = 5)]
-	V5(v5::Xcm<RuntimeCall>),
-}
-
-impl<C: Decode + GetDispatchInfo> IntoVersion for VersionedXcm<C> {
-	fn into_version(self, n: Version) -> Result<Self, ()> {
-		Ok(match n {
-			3 => Self::V3(self.try_into()?),
-			4 => Self::V4(self.try_into()?),
-			5 => Self::V5(self.try_into()?),
-			_ => return Err(()),
-		})
+versioned_type! {
+	pub enum VersionedXcm<RuntimeCall> {
+		#[codec(index = 3)]
+		V3(v3::Xcm<RuntimeCall>),
+		#[codec(index = 4)]
+		V4(v4::Xcm<RuntimeCall>),
+		#[codec(index = 5)]
+		V5(v5::Xcm<RuntimeCall>),
 	}
 }
 
-impl<C> IdentifyVersion for VersionedXcm<C> {
-	fn identify_version(&self) -> Version {
-		match self {
-			Self::V3(_) => v3::VERSION,
-			Self::V4(_) => v4::VERSION,
-			Self::V5(_) => v5::VERSION,
-		}
+/// TODO: workaround because `From<v4::Xcm>: v5::Xcm> is not satisfied
```

Comment:
I am not sure there is any

```suggestion
/// `versioned_type` doesn't implement this, so we implement manually.
```